File: modules/dns_enumeration.py
# -*- coding: utf-8 -*-

# MODULES AND/OR LIBRARIES
import socket
import concurrent.futures
import time
from modules.logging_config import debug, info, error

##############################

# RESOLVE SUBDOMAIN

##############################

def resolve_subdomain(subdomain):
    try:
        socket.gethostbyname(subdomain)
        return subdomain
    except:
        return None

##############################

# GET SUBDOMAINS

##############################

def get_subdomains(max_workers, target_domain_name, wordlist_path):
    start_time=time.time()
    subdomains = set()

    with open(wordlist_path, 'r') as f:
        subdomain_list = [line.strip() + "." + target_domain_name for line in f]

    print(f"[+] Starting DNS enumeration for: {target_domain_name}")
    print(f"[+] Wordlist: {wordlist_path}")
    print(f"[+] Max number of workers: {max_workers}")

    try:
        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
            futures = {executor.submit(resolve_subdomain, sub): sub for sub in subdomain_list}
            for future in concurrent.futures.as_completed(futures):
                result = future.result()
                if result:
                    subdomains.add(result)
    except KeyboardInterrupt:
        print("\n[!] Enumeration interrupted. Returning partial results.\n")
    finally:
        duration = time.time() - start_time
        print(f"\n[✓] Enumeration complete. Duration: {duration:.2f} seconds.")
        print(f"Total unique directories found: {len(subdomains)}\n")
        return subdomains

File: modules/dns_enumeration.go
package modules

import (
	"bufio"
	"context"
	"fmt"
	"log"
	"net"
	"os"
	"sync"
	"time"
)

// ##############################
//
// # CONSTANTS
//
// ##############################

// max concurrent DNS lookups
const maxConcurrentLookups = 50

// ##############################
//
// # SUBDOMAINS
//
// ##############################

func GetSubdomains(targetDomain string, wordlistPath string) {
	start := time.Now()
	fmt.Println("=== Subdomains ===")

	// Open wordlist file
	file, err := os.Open(wordlistPath)
	if err != nil {
		log.Println("Error opening wordlist:", err)
		return
	}
	defer file.Close()

	// Channel to limit concurrency
	sem := make(chan struct{}, maxConcurrentLookups)

	// WaitGroup for goroutines
	var wg sync.WaitGroup

	// Channel for results
	results := make(chan string)

	// Custom resolver with timeout
	resolver := &net.Resolver{}

	// Read wordlist line by line
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		subdomain := scanner.Text()
		if subdomain == "" {
			continue
		}

		wg.Add(1)
		sem <- struct{}{} // acquire token

		go func(sub string) {
			defer wg.Done()
			defer func() { <-sem }() // release token

			fullDomain := fmt.Sprintf("%s.%s", sub, targetDomain)

			// Context with timeout for lookup
			ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
			defer cancel()

			ips, err := resolver.LookupIP(ctx, "ip", fullDomain)
			if err != nil {
				// Optionally log DNS errors if needed, e.g.:
				// log.Printf("Lookup failed for %s: %v", fullDomain, err)
				return
			}
			if len(ips) > 0 {
				results <- fullDomain
			}
		}(subdomain)
	}

	// Close results channel once all lookups complete
	go func() {
		wg.Wait()
		close(results)
	}()

	// Print results as they come in
	for result := range results {
		fmt.Println(result)
	}

	if err := scanner.Err(); err != nil {
		log.Println("Error reading wordlist:", err)
	}

	duration := time.Since(start).Seconds()
	fmt.Printf("DNS enumeration completed in %.2f seconds.\n", duration)
}